### PR TITLE
[SPARK-48891][SS] Refactor StateSchemaCompatibilityChecker to unify all state schema formats

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Refactor StateSchemaCompatibilityChecker to unify all state schema formats

### Why are the changes needed?
Needed to integrate future changes around state data source reader and schema evolution and consolidate these changes

- Consolidates all state schema reader/writers in one place
- Consolidates all validation logic through the same API

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit tests

```
12:38:45.481 WARN org.apache.spark.sql.execution.streaming.state.StateSchemaCompatibilityCheckerSuite:

===== POSSIBLE THREAD LEAK IN SUITE o.a.s.sql.execution.streaming.state.StateSchemaCompatibilityCheckerSuite, threads: rpc-boss-3-1 (daemon=true), ForkJoinPool.commonPool-worker-3 (daemon=true), ForkJoinPool.commonPool-worker-2 (daemon=true), shuffle-boss-6-1 (daemon=true), ForkJoinPool.commonPool-worker-1 (daemon=true) =====
[info] Run completed in 12 seconds, 565 milliseconds.
[info] Total number of tests run: 30
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 30, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
cc - @ericm-db @jingz-db - PTAL, thx !

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 1]
```diff
@@ -30,19 +30,44 @@ import org.apache.spark.sql.execution.streaming.state.SchemaHelper.{SchemaReader
...
+    evolvedSchema: Boolean,
+    schemaPath: String
+)
+
+// Used to represent the schema of a column family in the state store
+case class StateStoreColFamilySchema(
+    colFamilyName: String,
+    keySchema: StructType,
+    valueSchema: StructType,
+    keyStateEncoderSpec: Option[KeyStateEncoderSpec] = None
```
Comment (ericm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:44]: do we not need userKeySchema?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala [hunk 2]
```diff
@@ -33,170 +28,164 @@ import org.apache.spark.util.Utils
...
+
+    override def read(inputStream: FSDataInputStream): List[StateStoreColFamilySchema] = {
+      val numEntries = inputStream.readInt()
+      (0 until numEntries).map { _ =>
+        val colFamilyName = inputStream.readUTF()
+        val keySchemaStr = readJsonSchema(inputStream)
+        val valueSchemaStr = readJsonSchema(inputStream)
+        StateStoreColFamilySchema(colFamilyName,
+          StructType.fromString(keySchemaStr),
+          StructType.fromString(valueSchemaStr))
```
Comment (ericm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala:98]: Don't you need to read/write the KeyStateEncoderSpec?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala [hunk 3]
```diff
@@ -33,170 +28,164 @@ import org.apache.spark.util.Utils
...
-      keySchema,
-      valueSchema,
-      KeyStateEncoderSpec.fromJson(keySchema, colFamilyMap("keyStateEncoderSpec")
-        .asInstanceOf[Map[String, Any]]),
-      colFamilyMap.get("userKeyEncoder").map(_.asInstanceOf[String]).map(StructType.fromString)
-    )
-  }
-}
+case class StateSchemaFormatV3(
+    stateStoreColFamilySchema: List[StateStoreColFamilySchema]
```
Comment (ericm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala:32]: Do we want to make this a V1? How will we deal with future versions of this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala [hunk 4]
```diff
@@ -18,58 +18,46 @@ package org.apache.spark.sql.execution.streaming
 
 import org.apache.spark.sql.Encoder
 import org.apache.spark.sql.execution.streaming.TransformWithStateKeyValueRowSchema.{COMPOSITE_KEY_ROW_SCHEMA, KEY_ROW_SCHEMA, VALUE_ROW_SCHEMA, VALUE_ROW_SCHEMA_WITH_TTL}
-import org.apache.spark.sql.execution.streaming.state.{ColumnFamilySchema, ColumnFamilySchemaV1, NoPrefixKeyStateEncoderSpec, PrefixKeyScanStateEncoderSpec}
+import org.apache.spark.sql.execution.streaming.state.{NoPrefixKeyStateEncoderSpec, PrefixKeyScanStateEncoderSpec, StateStoreColFamilySchema}
 
-trait ColumnFamilySchemaUtils {
-  def getValueStateSchema[T](stateName: String, hasTtl: Boolean): ColumnFamilySchema
+object StateSchemaUtils {
```
Comment (ericm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala:23]: Should we rename the file as well

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala (hunk 1 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala]: fixed this

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala (hunk 2 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala]: same here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala (hunk 3 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala:36]: Newer versions can just use a new state schema format version right ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala (hunk 4 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala [hunk 5]
```diff
@@ -235,14 +237,43 @@ class StateSchemaCompatibilityCheckerSuite extends SharedSparkSession {
...
+      StateStoreId(dir, opId, partitionId), queryId)
+    val runId = UUID.randomUUID()
+    val stateInfo = StatefulOperatorStateInfo(dir, runId, opId, 0, 200)
+    val storeColFamilySchema = List(
+      StateStoreColFamilySchema("test1", keySchema, valueSchema,
+        keyStateEncoderSpec = getKeyStateEncoderSpec(stateSchemaVersion, keySchema)),
+      StateStoreColFamilySchema("test2", longKeySchema, longValueSchema,
+        keyStateEncoderSpec = getKeyStateEncoderSpec(stateSchemaVersion, longKeySchema)),
+      StateStoreColFamilySchema("test3", keySchema65535Bytes, valueSchema65535Bytes,
+        keyStateEncoderSpec = getKeyStateEncoderSpec(stateSchemaVersion, keySchema65535Bytes)))
```
Comment (ericm-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala:265]: Can we add one with userKeySchema specified?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StreamStreamJoinStateHelper.scala [hunk 6]
```diff
@@ -68,11 +68,11 @@ object StreamStreamJoinStateHelper {
     val newHadoopConf = session.sessionState.newHadoopConf()
 
     val manager = new StateSchemaCompatibilityChecker(providerIdForKeyToNumValues, newHadoopConf)
-    val (keySchema, _) = manager.readSchemaFile()
+    val keySchema = manager.readSchemaFile().head.keySchema
 
     val manager2 = new StateSchemaCompatibilityChecker(providerIdForKeyWithIndexToValue,
```
Comment (ericm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StreamStreamJoinStateHelper.scala:73]: can you add a comment on what this is doing? Why are we reading the keySchema from one manager and valueSchema from another?

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StreamStreamJoinStateHelper.scala:75]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala (hunk 5 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala]: Done

**Review by ericm-db — APPROVED** [review]
LGTM

**Comment by anishshri-db** [comment]
@HeartSaVioR - could you PTAL ? Thx !

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateStoreColumnFamilySchemaUtils.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateStoreColumnFamilySchemaUtils.scala:1]: Will this change be rebased with #47257? I'm not sure what would be the final outcome. I'll defer the review of this file (and related change) after rebase.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala [hunk 7]
```diff
@@ -33,170 +32,190 @@ import org.apache.spark.util.Utils
...
+        val keyEncoderSpecStr = readJsonSchema(inputStream)
+        val colFamilyMap = JsonMethods.parse(keyEncoderSpecStr).extract[Map[String, Any]]
+        val encoderSpec = KeyStateEncoderSpec.fromJson(keySchema, colFamilyMap)
+
+        // read the user key encoder spec if provided
+        val userKeyEncoderSchemaStr = readJsonSchema(inputStream)
+        val userKeyEncoderSchema = Try(StructType.fromString(userKeyEncoderSchemaStr)).toOption
+
+        StateStoreColFamilySchema(colFamilyName,
+          keySchema,
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala:114]: nit: inline? just to be consistent with below

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala [hunk 8]
```diff
@@ -235,14 +237,46 @@ class StateSchemaCompatibilityCheckerSuite extends SharedSparkSession {
...
+  test("checking for compatibility with schema version 3") {
+    val stateSchemaVersion = 3
+    val dir = newDir()
+    val queryId = UUID.randomUUID()
+    val providerId = StateStoreProviderId(
+      StateStoreId(dir, opId, partitionId), queryId)
+    val runId = UUID.randomUUID()
+    val stateInfo = StatefulOperatorStateInfo(dir, runId, opId, 0, 200)
+    val storeColFamilySchema = List(
+      StateStoreColFamilySchema("test1", keySchema, valueSchema,
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala:260]: Why not test with various encoder specs here given we are testing version 3 specifically?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 9]
```diff
@@ -140,8 +147,32 @@ object StateSchemaCompatibilityChecker extends Logging {
     }
   }
 
-  private def schemasCompatible(storedSchema: StructType, schema: StructType): Boolean =
-    DataType.equalsIgnoreNameAndCompatibleNullability(schema, storedSchema)
+  def validateAndMaybeEvolveStateSchema(
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:156]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 10]
```diff
@@ -30,19 +30,45 @@ import org.apache.spark.sql.execution.streaming.state.SchemaHelper.{SchemaReader
...
+    valueSchema: StructType,
+    keyStateEncoderSpec: Option[KeyStateEncoderSpec] = None,
+    userKeyEncoderSchema: Option[StructType] = None
+)
+
 class StateSchemaCompatibilityChecker(
     providerId: StateStoreProviderId,
-    hadoopConf: Configuration) extends Logging {
+    hadoopConf: Configuration,
+    stateSchemaVersion: Int = 2,
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateStoreColumnFamilySchemaUtils.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateStoreColumnFamilySchemaUtils.scala:1]: Yes - will rebase after Jing's changes are merged

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala (hunk 7 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala:114]: @HeartSaVioR - not sure what you mean here ? what would you like to change here ?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala (hunk 8 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala (hunk 7 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala:114]: We clarified the usage - keySchema is referred in above.

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala:114]: Discussed offline - we need the `keySchema` in the line above (at 107).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateStoreColumnFamilySchemaUtils.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateStoreColumnFamilySchemaUtils.scala:1]: Done

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**